File: C#/src/core/StatusMessage.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

using System.IO;
using System.Xml;

namespace RobotEngine.Messaging{
	public class StatusMessage
	{
		/// <summary>
		/// The task identifier.
		/// </summary>
		private string _taskID;
		
		/// <summary>
		/// The task's current status.
		/// </summary>
		private string _status;
		
		/// <summary>
		/// The optional set of status details.
		/// </summary>
		private Dictionary<string, string> _statusDetails;
		
		
		/// <summary>
		/// Creates a StatusMessage with the given task ID and status type.
		/// </summary>
		/// <param name="taskID">the task identifier</param>
		/// <param name="status">the task's current status</param>
		public StatusMessage(string taskID, string status)
		{
			if((taskID==null)||(status==null))
				throw new System.ArgumentNullException("StatusMessage requires at least a task ID and status label");
			
			_taskID = taskID;
			_status = status;
			_statusDetails = new Dictionary<string, string>();
		}

		/// <summary>
		/// Creates a StatusMessage by parsing its XML representation.
		/// </summary>
		/// <param name="msgString">the XML string containing the message data</param>
		public StatusMessage(string msgString)
		{
			XmlReader reader = XmlReader.Create(new StringReader(msgString));
			
			XmlDocument doc = new XmlDocument();
			doc.Load(reader);
			
			XmlNodeList statusNodes = doc.GetElementsByTagName("status");
			if(statusNodes.Count==0)
				throw new InvalidDataException("message string does not contain any <status> element!");
			
			_statusDetails = new Dictionary<string, string>();
			
			var iter = statusNodes[0].Attributes.GetEnumerator();
			while(iter.MoveNext())
			{
				XmlAttribute attr = (XmlAttribute)iter.Current;
				if(attr.Name.Equals("task"))
				{
					_taskID = attr.Value;
				}
				else if(attr.Name.Equals("status"))
				{
					_status = attr.Value;
				}
				else{
					//arbitrary command parameter
					_statusDetails.Add(attr.Name, attr.Value);
				}
			}
			
			//validate the message
			if((_taskID==null)||(_status==null))
				throw new System.ArgumentNullException("StatusMessage requires at least a task ID and status label");
		}


		public XmlDocument ToDocument(){
			XmlDocument doc = new XmlDocument();
			XmlElement statElem = doc.CreateElement("status");
			
			statElem.SetAttribute("task", _taskID);
			statElem.SetAttribute("status", _status);
			
			var iter = _statusDetails.GetEnumerator();
			while(iter.MoveNext())
			{
				KeyValuePair<string,string> entry = iter.Current;
				
				statElem.SetAttribute(entry.Key, entry.Value);
			}
			
			doc.AppendChild(statElem);
			
			return doc;
		}
		
		
		public override string ToString()
		{
			XmlDocument doc = ToDocument();
			
			StringWriter writer = new StringWriter();
			XmlWriter xmlWriter = XmlWriter.Create(writer);
			
			doc.WriteTo(xmlWriter);
			xmlWriter.Flush();
			
			return writer.ToString();
		}
		
		public void AddDetail(string detailName, string detailValue)
		{
			_statusDetails.Add(detailName, detailValue);
		}
		
		public void RemoveDetail(string detailName)
		{
			_statusDetails.Remove(detailName);
		}
		
		public string GetTaskID(){
			return _taskID;
		}
		
		public string GetStatus(){
			return _status;
		}
		
		public Dictionary<string,string> GetStatusDetails(){
			return _statusDetails;
		}
		
		public string GetDetail(string detailName){
			string value;
			bool success = _statusDetails.TryGetValue(detailName, out value);
			if(success)
				return value;
			else return null;
		}
	}
}

File: Java/src/de/kmj/robots/messaging/StatusMessage.java
package de.kmj.robots.messaging;

import static de.kmj.robots.messaging.XMLMessage.cLogger;
import java.util.Map.Entry;
import java.util.Set;
import java.util.TreeMap;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.StringWriter;
import java.io.UnsupportedEncodingException;
import java.util.logging.Level;
import java.util.logging.Logger;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.TransformerFactoryConfigurationError;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

/**
 * The data structure for messages sent from the robot to the control
 * application.
 *
 * Every status message contains at least a task identifier and a status label
 * which can be interpreted by the conrol application. Additionally, it can hold
 * an arbitrary set of details about the status change.
 *
 * The message can be converted to and from an XML string which is used for
 * transmitting it over network sockets.
 *
 * @author Kathrin Janowski
 */
public class StatusMessage extends XMLMessage{

    /**
     * The task identifier.
     */
    private String mTaskID;

    /**
     * The task's current status.
     */
    private String mStatus;

    /**
     * The optional set of status details.
     */
    private TreeMap<String, String> mStatusDetails;

    /**
     * Creates a StatusMessage with the given task ID and status label.
     *
     * @param taskID the task identifier
     * @param status the task's current status
     * @throws IllegalArgumentException if the taskID or the command type is
     * null
     */
    public StatusMessage(String taskID, String status)
            throws IllegalArgumentException {
        
        super();
        
        // validate the message
        if (taskID == null || status == null) {
            throw new IllegalArgumentException(
                    "StatusMessage requires at least a task ID and status label");
        }

        mTaskID = taskID;
        mStatus = status;

        mStatusDetails = new TreeMap<String, String>();
    }

    /**
     * Creates a StatusMessage by parsing its XML representation.
     *
     * @param messageStr the XML string containing the message data
     * @throws IllegalArgumentException if the taskID or the command type is
     * missing
     */
    public StatusMessage(String messageStr) throws IllegalArgumentException {
        
        super();
        parseDocument(messageStr);

        // validate the message
        if (mTaskID == null || mStatus == null) {
            throw new IllegalArgumentException(
                    "StatusMessage requires at least a task ID and status label");
        }
    }

    /**
     * Creates a StatusMessage by parsing an individual XML node.
     *
     * @param messageNode the XML node containing the message data
     * @throws IllegalArgumentException if the taskID or the status label is
     * missing
     */
    public StatusMessage(Node messageNode){
        super();
        createDocument();
        mDocument.appendChild(messageNode);
        parseMessageContent();
    }
    
    
    @Override
    protected void parseMessageContent()
    {
        if(mDocument==null)
        {
            cLogger.log(Level.SEVERE, "Can't parse status content: The document was not created yet!");
            return;
        }
        
        
        NodeList statusElements = mDocument.getElementsByTagName("status");
        if(statusElements.getLength()==0)
        {
            cLogger.log(Level.SEVERE, "No <status> element in the XML message!");
            return;
        }
        
        //------------------------------------------------------------------
        // parse all attributes
        //------------------------------------------------------------------
        Element statusNode = (Element)statusElements.item(0);
        mStatusDetails = new TreeMap<String, String>();

        NamedNodeMap attributes = statusNode.getAttributes();
        int attrCount = attributes.getLength();
        for (int i = 0; i < attrCount; i++) {
            Node attr = attributes.item(i);
            String attrName = attr.getNodeName();
            String attrValue = attr.getNodeValue();

            if (attrName.equals("task")) {
                mTaskID = attrValue;
            } else if (attrName.equals("status")) {
                mStatus = attrValue;
            } else //arbitrary status details
            {
                mStatusDetails.put(attrName, attrValue);
            }
        }
        
        // validate the message
        if (mTaskID == null || mStatus == null) {
            throw new IllegalArgumentException(
                    "StatusMessage requires at least a task ID and status label");
        }
    }

    
    
    @Override
    public Element createMessageElement(Document doc){
        Element statusElem = doc.createElement("status");
        statusElem.setAttribute("task", mTaskID);
        statusElem.setAttribute("status", mStatus);
            
        Set<Entry<String, String>> params = mStatusDetails.entrySet();
        for (Entry<String, String> param : params)
            statusElem.setAttribute(param.getKey(), param.getValue());
        
        return statusElem;
    }



   
    /*==========================================================================
     *  getters
     *==========================================================================*/
    public String getTaskID() {
        return mTaskID;
    }

    public String getStatus() {
        return mStatus;
    }

    public TreeMap<String, String> getStatusDetails() {
        return mStatusDetails;
    }

    public String getDetail(String detailName) {
        return mStatusDetails.get(detailName);
    }
    
        /**
     * Adds the given status detail. If the entry exists, its value is replaced.
     *
     * @param detailName the detail name
     * @param detailValue the detail value
     */
    public void addDetail(String detailName, String detailValue) {
        mStatusDetails.put(detailName, detailValue);
    }

    /**
     * Removes the given status detail.
     *
     * @param detailName the parameter name
     */
    public void removeDetail(String detailName) {
        mStatusDetails.remove(detailName);
    }

}
